illumos-utils/src/zfs.rs

```diff
@@ -493,6 +506,196 @@ pub struct DatasetEnsureArgs<'a> {
     pub additional_options: Option<Vec<String>>,
 }
 
+// If this dataset will have a well-defined mountpoint, we ensure an immutable directory exists
+// here.
+//
+// This is intended to mitigate issues like:
+// - https://github.com/oxidecomputer/omicron/issues/7874
+// - https://github.com/oxidecomputer/omicron/issues/4203
+//
+// Suppose we have a dataset, named "crypt/debug", within a parent dataset named "crypt".
+//
+// Suppose "crypt" is mounted at "/storage/crypt", and we want to mount "crypt/debug" as
+// "/storage/crypt/debug".
+//
+// When we create the debug dataset, we are creating a directory within the "crypt" dataset, and
+// mounting our new dataset on top of it. Without modification, this presents a threat: the
+// "/storage/crypt/debug" directory is a regular directory, and if the underlying dataset is
+// unmounted (explicitly, or due to a reboot) data could be placed UNDERNEATH the mount point.
+//
+// To mitigate this issue, we create the mountpoint ahead-of-time, and set the immutable
+// property on it. This prevents the mountpoint from being used as anything other than a
+// mountpoint.
+//
+// NOTE: This must only be called on mountpoints for datasets which ARE NOT MOUNTED.
+fn ensure_empty_immutable_mountpoint(
+    mountpoint: &Utf8Path,
+) -> Result<(), EnsureDatasetErrorRaw> {
+    if !mountpoint
+        .try_exists()
+        .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?
+    {
+        std::fs::create_dir_all(mountpoint)
+            .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?;
+
+        // We still fall-through to the logic below, even though we just created
+        // this directory. It's technically possible someone put a file inside
+        // it.
+    }
+
+    // "making a directory empty" and "making a directory immutable" cannot
+    // be atomic together. Here's how we cope:
+    //
+    // - If necessary, we make the directory mutable, and try to clear it out.
+    // - We then re-set the directory to immutable.
+    //
+    // This means that if someone adds a file to our mountpoint after
+    // "emptying it out" but before "making it immutable", we make
+    // another attempt at making it empty.
+    loop {
+        let immutablity = is_directory_immutable(mountpoint)?;
```

Comment:
Not all filesystem support immutable attributes. We read here before attempting to set any immutability, and don't bother trying to make things immutable if the attribute isn't supported.

This is the case on the "helios/deploy" job, which uses file-backed synthetic disks.